### PR TITLE
MCR-2315 Common methods for {base,type,project} CLI command variants

### PR DESCRIPTION
**Pull request by OPNA2608**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2315).

Includes trivial migrations of various commands to these new helper methods.

### CODE REVIEW

**Comment by yagee-de** [comment]
Please add Javadoc to public methods of MCRCommandUtils

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 262223cbb98f842892ceac83ef8ecef767755ee8! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6283-pull_request-1117

**Comment by yagee-de** [comment]
It is already in a good shape, but I think it does not use Java-Features well enough.
There should be helper methods to generate Stream instances for different types, baseIds and from-to-intervalls, and what ever is necessary to get those streams generated. Then we should use `Function<String, String>` to map single values to a command. That would help to get rid of `String... otherValues` as this is moved to a lambda expression.


```java

//example now
return MCRCommandUtils.runCommandOnBaseId("repair metadata search of ID {0}", baseID);
//example later
return MCRCommandUtils.getIdsForBaseId(base)
    .map(id -> "repair metadata search of ID "+id)
    .collect(Collectors.toList());
```

This will make the calls more readable as the command is clear instead of splitting it over several parameters, if `otherValues` is used.

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 80b50e66131ed4526649e3c52a74a6943080ffb1! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6338-pull_request-1117

**Comment by OPNA2608** [comment]
@yagee-de Could you take another look please? I dropped the `runCommandOn` methods in favour of using the `getIdsFor` ones everywhere like in your code example.

File: mycore-pi/src/main/java/org/mycore/pi/doi/cli/MCRDOICommands.java [hunk 1]
```diff
@@ -177,19 +177,14 @@ private static MCRObjectID getObjectID(URI uri) {
...
-                    boolean isTestDOI = doi.getPrefix().equals(MCRDigitalObjectIdentifier.TEST_DOI_PREFIX);
-                    return !isTestDOI;
-                })
-                .map(MCRDigitalObjectIdentifier::asString)
-                .map(doiStr -> new MessageFormat(REPAIR_MEDIALIST_OF_0_AND_SERVICE_1, Locale.ROOT).format(
-                    new Object[] { doiStr, serviceID }))
+            return dataciteClient.getDOIList().stream().filter(doi -> {
+                boolean isTestDOI = doi.getPrefix().equals(MCRDigitalObjectIdentifier.TEST_DOI_PREFIX);
+                return !isTestDOI;
+            }).map(MCRDigitalObjectIdentifier::asString).map(id -> syntaxMF.format(new Object[] { id }))
```
Comment (yagee-de) [inline mycore-pi/src/main/java/org/mycore/pi/doi/cli/MCRDOICommands.java]: Does not seem to provide the right amount of parameters. Why the change?

File: mycore-mods/src/main/java/org/mycore/mods/MCRMODSCommands.java [hunk 2]
```diff
@@ -82,11 +82,9 @@
...
             return null;
         }
-        return Arrays.stream(list)
-            .filter(file -> file.endsWith(".xml"))
-            .map(file -> String.format(Locale.ENGLISH,
-                "load mods document from file %s for project %s",
-                new File(dir, file).getAbsolutePath(), projectID))
+        return Arrays.stream(list).parallel().filter(file -> file.endsWith(".xml"))
+            .map(file -> new File(dir, file).getAbsolutePath())
+            .map(file -> "load mods document from file " + file + " for project " + projectID)
```
Comment (yagee-de) [inline mycore-mods/src/main/java/org/mycore/mods/MCRMODSCommands.java]: Why the change?

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            throw new MCRUsageException(
+                "MCR.Metadata.Type." + type + " not set!");
+        }
+        List<String> idList;
+        try {
+            idList = MCRXMLMetadataManager.instance().listIDsOfType(type);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new MCRUsageException("Requested type " + type + " not found!", e);
+        }
+        if (idList.size() == 0) {
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: use isEmpty()

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        if (base == null || base.length() == 0) {
+            throw new MCRUsageException("Base ID ({project}_{type}) required to enumerate IDs!");
+        }
+        List<String> idList;
+        try {
+            idList = MCRXMLMetadataManager.instance().listIDsForBase(base);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new MCRUsageException("Requested base " + base + " not found!", e);
+        }
+        if (idList.size() == 0) {
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: use isEmpty()

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        }
+        int fromID = from.getNumberAsInteger();
+        int toID = to.getNumberAsInteger();
+        int lowerBound = fromID < toID ? fromID : toID;
+        int upperBound = fromID < toID ? toID : fromID;
+        List<String> idList = IntStream.rangeClosed(lowerBound, upperBound).boxed().parallel()
+            .map(n -> MCRObjectID.formatID(fromBase, n))
+            .filter(id -> MCRMetadataManager.exists(MCRObjectID.getInstance(id)))
+            .collect(Collectors.toCollection(ArrayList::new));
+        if (idList.size() == 0) {
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: use isEmpty()

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        }
+        if (MCRConfiguration2.getString("MCR.Metadata.Type." + type).isEmpty()) {
+            throw new MCRUsageException(
+                "MCR.Metadata.Type." + type + " not set!");
+        }
+        List<String> idList;
+        try {
+            idList = MCRXMLMetadataManager.instance().listIDsOfType(type);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new MCRUsageException("Requested type " + type + " not found!", e);
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: Do not throw an exception here! Never ever catch RuntimeException if you cannot handle them.

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            throw new MCRUsageException("Start-ID required to enumerate IDs!");
+        }
+        if (endId == null || endId.length() == 0) {
+            throw new MCRUsageException("End-ID required to enumerate IDs!");
+        }
+        MCRObjectID from, to;
+        try {
+            from = MCRObjectID.getInstance(startId);
+        } catch (MCRException e) {
+            throw new MCRUsageException("Start-ID " + startId + " is invalid!", e);
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: Do not catch this exception here.

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        }
+        try {
+            to = MCRObjectID.getInstance(endId);
+        } catch (MCRException e) {
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: ..or here

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        String toBase = to.getBase();
+        if (!fromBase.equals(toBase)) {
+            throw new MCRUsageException(
+                startId + " and " + endId + " have different base IDs (" + fromBase + " and " + toBase
+                    + "), same base required to enumerate IDs!");
+        }
+        try {
+            MCRXMLMetadataManager.instance().getStore(fromBase, true);
+        } catch (MCRPersistenceException e) {
+            throw new MCRUsageException(
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: Do not catch a RuntimeException if you cannot handle it.

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+     *     Not thrown if type exists but has no values.
+     */
+    public static final Stream<String> getIdsForType(final String type) throws MCRUsageException {
+        if (type == null || type.length() == 0) {
+            throw new MCRUsageException("Type required to enumerate IDs!");
+        }
+        if (MCRConfiguration2.getString("MCR.Metadata.Type." + type).isEmpty()) {
+            throw new MCRUsageException(
+                "MCR.Metadata.Type." + type + " not set!");
+        }
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: Only require a property, if you use it. Code you call later may change and requires to check the input data. Do not duplicate checks!

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+     * 
+     * @param base
+     *     The base to look in the store for.
+     * @return A (parallel) stream with found IDs. (may be empty if none were found)
+     * @throws MCRUsageException
+     *     If no base was given or the base could not be found.
+     *     Not thrown if the base exists but has no values.
+     */
+    public static final Stream<String> getIdsForBaseId(final String base) throws MCRUsageException {
+        if (base == null || base.length() == 0) {
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: If you want to check a valid "base" here check for "_" or use MCRObjectID.getIDParts()

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            throw new MCRUsageException("Base ID ({project}_{type}) required to enumerate IDs!");
+        }
+        List<String> idList;
+        try {
+            idList = MCRXMLMetadataManager.instance().listIDsForBase(base);
+        } catch (ArrayIndexOutOfBoundsException e) {
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: do not catch RuntimeException here, code may change and will leave this code dead here.

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-mods/src/main/java/org/mycore/mods/MCRMODSCommands.java (hunk 2 above)
Comment (OPNA2608) [reply mycore-mods/src/main/java/org/mycore/mods/MCRMODSCommands.java]: Fallout from the `runCommandOn` removal, will drop the change.

File: mycore-pi/src/main/java/org/mycore/pi/doi/cli/MCRDOICommands.java (hunk 1 above)
Comment (OPNA2608) [reply mycore-pi/src/main/java/org/mycore/pi/doi/cli/MCRDOICommands.java]: Same fallout, will drop.

**Comment by yagee-de** [comment]
It's hard to start from scratch every review. Please consider adding single commits in the future.

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public static final Stream<String> getIdsForType(final String type) throws MCRUsageException {
+        if (type == null || type.length() == 0) {
+            throw new MCRUsageException("Type required to enumerate IDs!");
+        }
+        List<String> idList = MCRXMLMetadataManager.instance().listIDsOfType(type);
+        if (idList.isEmpty()) {
+            LOGGER.warn("No IDs found for type {}.", type);
+        }
+        return idList.parallelStream();
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: use stream() not parallelStream()

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+     */
+    public static final Stream<String> getIdsForBaseId(final String base) throws MCRUsageException {
+        if (MCRObjectID.getIDParts(base).length == 2) {
+            throw new MCRUsageException("Base ID ({project}_{type}) required to enumerate IDs!");
+        }
+        List<String> idList = MCRXMLMetadataManager.instance().listIDsForBase(base);
+        if (idList.isEmpty()) {
+            LOGGER.warn("No IDs found for base {}.", base);
+        }
+        return idList.parallelStream();
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: use stream() not parallelStream()

File: mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+        int lowerBound = fromID < toID ? fromID : toID;
+        int upperBound = fromID < toID ? toID : fromID;
+        List<String> idList = IntStream.rangeClosed(lowerBound, upperBound).boxed().parallel()
+            .map(n -> MCRObjectID.formatID(fromBase, n))
+            .filter(id -> MCRMetadataManager.exists(MCRObjectID.getInstance(id)))
+            .collect(Collectors.toCollection(ArrayList::new));
+        if (idList.isEmpty()) {
+            LOGGER.warn("No IDs found in range [{} -> {}].", from, to);
+        }
+        return idList.parallelStream();
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/frontend/cli/MCRCommandUtils.java]: use stream() not parallelStream()

**Review by yagee-de — CHANGES_REQUESTED** [review]
Do not return parallelStreams by default. Will cause issues with session management and database connections.

**Review by yagee-de — APPROVED** [review]

**Pull request merged**